File: Swift-C++-Testsuite/test/protocols/Swift/ConditionallyConformingToProtocol.swift
@main
struct MyMain {

    static func main() {
        let d12 = Dice(sides: 12, generator: LinearCongruentialGenerator())
        let d6 = Dice(sides: 6, generator: LinearCongruentialGenerator())
        let myDice = [d6, d12]
        print(myDice.textualDescription)
    }

}
class Dice {
    let sides: Int
    let generator: RandomNumberGenerator
    init(sides: Int, generator: RandomNumberGenerator) {
        self.sides = sides
        self.generator = generator
    }
    func roll() -> Int {
        return Int(generator.random() * Double(sides)) + 1
    }
}
extension Dice: TextRepresentable {
    var textualDescription: String {
        return "A \(sides)-sided dice"
    }
}
protocol TextRepresentable {
    var textualDescription: String { get }
}
extension Array: TextRepresentable where Element: TextRepresentable {
    var textualDescription: String {
        let itemsAsText = self.map { $0.textualDescription }
        return "[" + itemsAsText.joined(separator: ", ") + "]"
    }
}

class LinearCongruentialGenerator: RandomNumberGenerator {
    var lastRandom = 42.0
    let m = 139968.0
    let a = 3877.0
    let c = 29573.0
    func random() -> Double {
        lastRandom =
            ((lastRandom * a + c)
                .truncatingRemainder(dividingBy: m))
        return lastRandom / m
    }
}

protocol RandomNumberGenerator {
    func random() -> Double
}

File: Swift-C++-Testsuite/test/protocols/C++/ConditionallyConformingToProtocol.cpp

 #include <iostream>
 #include <string>
 #include <vector>
 #include <math.h>
 
 class RandomNumberGenerator {
	public:
		virtual double random() = 0;
 };

 class Dice {
	 int sides;
	 RandomNumberGenerator *generator;
 public:
	 Dice(int sides, RandomNumberGenerator *generator) : sides(sides), generator(generator) {
 
	 }
	 int roll() {
		 return int(generator->random() * sides) + 1;
	 }
	 virtual std::string textualDescription() {
		 return "A " + std::to_string(sides) + "-sided dice";
	 }
 };
 
 class LinearCongruentialGenerator : public RandomNumberGenerator {
	 double lastRandom = 42.0;
	 double m = 139968.0;
	 double a = 3877.0;
	 double c = 29573.0;
 public:
	 double random() {
		 lastRandom = fmod((lastRandom * a + c), m);
		 return lastRandom / m;
	 }
 };
 
 template<typename T>
 class TextRepresentable {
 public:
	 virtual std::string textualDescription() = 0;
 };
 
 template<typename T>
 class Vector : public TextRepresentable<T> {
	 std::vector<T> myDice;
 public:
	 Vector(std::vector<T> myDice) : myDice(myDice) {
 
	 }
	 std::string textualDescription() {
		 std::string res = "[";
		 for (int i = 0; i < myDice.size(); i++) {
			 res += myDice[i]->textualDescription();
			 if (i != myDice.size() - 1) {
				 res += ", ";
			 }
		 }
		 res += "]";
		 return res;
	 }
 };
 
 int main() {
	 LinearCongruentialGenerator *src = new LinearCongruentialGenerator();
	 Dice *d12 = new Dice(12, src);
	 Dice *d6 = new Dice(6, src);
	 std::vector<Dice*> myDice = { d6, d12 };
	 Vector<Dice*> vec(myDice);
	 std::cout << vec.textualDescription();
	 delete src;
	 delete d12;
	 delete d6;
 }